### ISSUE TITLE
Proposal for versioned samples images

### ISSUE THREAD

**Issue by lbussell** [dotnet/dotnet-docker#4932]
### Describe the Problem

Some customers have expressed the desire for accessing old versions of samples in the dotnet/samples repo. We want .NET 8.0 samples to be compatible with older samples versions but still want to publish new samples with the latest .NET Docker features/best practices.

### Proposal

- Keep both `dotnetapp` and `aspnetapp` samples in the same `dotnet/samples` repo. This promotes visibility between different types of samples and also allows us to introduce more types of samples in the future with less overhead.
- Keep the floating tags published using traditional, rootful containers with a shell.
  - Consider moving to non-root or distroless for a future release.
- Add a new tag with a version suffix to all samples. This can be undocumented if necessary.
  - The new full tag structure would be `(sample app name)-(optional sample variant)-(.NET major version)`.
  - For each versioned tag, there should also be a floating versionless tag that we will point customers to. `(sample name)-(variant)`

#### Tags

I have provided several examples of the proposed tagging scheme below. The list is not exhaustive, the idea is just to provide some examples of how we could publish different types of samples with this naming scheme.

| Old Tags | New Tags | OS |
|---|---|---|
| `aspnetapp` | `aspnetapp`, `aspnetapp-8.0` | Alpine 3.18 |
| Not published | `aspnetapp-composite`, `aspnetapp-composite-8.0` | Jammy Chiseled |
| Not published | `aspnetapp-aot`, `aspnetapp-aot-8.0` | Jammy Chiseled |
| Not published | `aspnetapp-chiseled`, `aspnetapp-chiseled-8.0`<br>OR `aspnetapp-distroless`, `aspnetapp-distroless-8.0`... | Jammy Chiseled |
| | etc. | |
| `dotnetapp` | `dotnetapp`, `dotnetapp-8.0` | Alpine 3.18 |
| Not published | `dotnetapp-chiseled`, `dotnetapp-chiseled-8.0` | Jammy Chiseled |
| Not published | `dotnetapp-aot`, `dotnetapp-aot-8.0` | Jammy Chiseled |

**Comment by lbussell** [comment]
Updated the tag listings to include OS versions.

**Comment by lbussell** [comment]
Here is a greatly simplified proposal based on working with @richlander:


| Old Tags | New Tags | OS |
|---|---|---|
| `aspnetapp`, `aspnetapp-alpine-<arch>` | `aspnetapp`, `aspnetapp-8.0`, `aspnetapp-8.0-alpine-<arch>` | Alpine |
| | `aspnetapp-chiseled`, `aspnetapp-chiseled-8.0`, `aspnetapp-chiseled-8.0-<arch>` | Ubuntu Chiseled |
| `dotnetapp`, `dotnetapp-alpine-<arch>` | `dotnetapp`, `dotnetapp-8.0`, `dotnetapp-8.0-alpine3.18-<arch>` | Alpine |
| | `dotnetapp-chiseled`, `dotnetapp-chiseled-8.0`, `dotnetapp-8.0-chiseled-<arch>` | Ubuntu Chiseled |

The intention of having a versioned multi-platform tag (e.g. `aspnetapp-8.0`) is so that in the future, it would be easy to compare samples across .NET versions to see startup time and/or size differences, using a simple and accessible tag. I am not super attached to that idea so I could be convinced to remove it.

**Comment by lbussell** [comment]
[Triage] After some discussion, we think it would be best to leave Alpine as default so that users can inspect the image with a shell. The Chiseled samples should not be shipped with the addition of a shell since that is a design goal for the Chiseled images and we want to accurately portray the final product that a user might get when using them.

**Comment by richlander** [comment]
Agree. I key change we need is to set the `USER` in the `aspnetapp` sample.

**Comment by lbussell** [comment]
> Agree. I key change we need is to set the `USER` in the `aspnetapp` sample.

[Triage] We should leave the user as `root` in .NET 8 and switch it to non-root in .NET 9.

**Comment by lbussell** [comment]
Versioned samples and Ubuntu Chiseled are now available with https://github.com/dotnet/dotnet-docker/pull/5185

https://mcr.microsoft.com/en-us/product/dotnet/samples/about

`docker pull mcr.microsoft.com/dotnet/samples:aspnetapp-chiseled-8.0`

![Image](https://github.com/dotnet/dotnet-docker/assets/36081148/3006cffa-ad6a-4ae2-b3e6-9c5e57d8e5f1)
